model/PermissionsService.php

```diff
@@ -45,6 +45,9 @@ class PermissionsService
     /** @var EventManager */
     private $eventManager;
 
+    /** @var bool */
+    private $isRecursive = true;
```

Comment:
You are changing the state of the service impacting various methods and it is not clear for the client of this class. It can cause many issues. Example: 

- if I call `savePermissions` with `isRecursive = false`, the it will never be possible to call `saveResourcePermissions` recursively.
- if you want to call `saveResourcePermissions` recursively is not possible anymore.

I would either create splicity public methods like `saveResourcePermissions` and `saveResourcePermissionsRecursively`, so it is clear what they do.

This service is already becoming too big, I would instead create one more specific (i.e `ResourcePermissionService`) and add those methods there. That is the problem of too generic names in classes, you can justify SRP as long as the name is generic. =)